gflownet/envs/crystals/crystal.py

```diff
@@ -0,0 +1,488 @@
+        )
+
+        # start and end indices of individual submasks
+        self.composition_mask_start = 0
+        self.composition_mask_end = self.composition_mask_start + len(
+            self.composition.action_space
+        )
+        self.space_group_mask_start = self.composition_mask_end
+        self.space_group_mask_end = self.space_group_mask_start + len(
+            self.space_group.action_space
+        )
+        self.lattice_parameters_mask_start = self.space_group_mask_end
+        self.lattice_parameters_mask_end = self.lattice_parameters_mask_start + len(
+            self.lattice_parameters.action_space
+        )
+
+        self.eos = self.lattice_parameters.eos
+        self.stage = Stage.COMPOSITION
+        self.max_action_length = max(
+            max(len(a) for a in self.composition.action_space),
+            max(len(a) for a in self.space_group.action_space),
+            max(len(a) for a in self.lattice_parameters.action_space),
+        )
+        self.done = False
+
+        # Conversions
+        self.state2proxy = self.state2oracle
+        self.statebatch2proxy = self.statebatch2oracle
+        self.statetorch2proxy = self.statetorch2oracle
+
+        super().__init__(**kwargs)
+
+    def _set_lattice_parameters(self):
+        """
+        Sets LatticeParameters conditioned on the lattice system derived from the SpaceGroup.
+        """
+        crystal_system = self.space_group.get_crystal_system()
+
+        if crystal_system is None:
+            raise ValueError(
+                "Cannot set lattice parameters without crystal system determined in the space group."
+            )
+
+        self.lattice_parameters = LatticeParameters(
+            lattice_system=self.space_group.lattice_system,
+            **self.lattice_parameters_kwargs,
+        )
+
+    def _pad_action(self, action: Tuple[int], stage: Stage) -> Tuple[int]:
+        """
+        Pads action such that all actions, regardless of the underlying environment, have
+        the same length. Required due to the fact that action space has to be convertable to
+        a tensor.
+        """
+        return action + (stage.value,) * (self.max_action_length - len(action))
+
+    def _pad_action_space(
+        self, action_space: List[Tuple[int]], stage: Stage
+    ) -> List[Tuple[int]]:
+        return [self._pad_action(a, stage) for a in action_space]
+
+    def _depad_action(self, action: Tuple[int], stage: Stage) -> Tuple[int]:
+        """
+        Reverses padding operation, such that the resulting action can be passed to the
+        underlying environment.
+        """
+        if stage == Stage.COMPOSITION:
+            dim = max(len(a) for a in self.composition.action_space)
+        elif stage == Stage.SPACE_GROUP:
+            dim = max(len(a) for a in self.space_group.action_space)
+        elif stage == Stage.LATTICE_PARAMETERS:
+            dim = max(len(a) for a in self.lattice_parameters.action_space)
```

Comment:
These for loops can be avoided by storing the max values, right? This may not be negligible because the action spaces are large and this will be called many many times (each call to `step()` for example)

Reply:
Actually, now I think about it, the length of the actions should be the same for a given environment. At least this is the case for the space group environment. I started coding things this way since I decided that at some point I want to port this to all-torch-tensors.

Reply:
Good point about computational cost, I changed it to be computed only once.

Regarding same action length, I know, and I believe this is the case right now, but I think it's worth checking it once anyway to prevent issues with any possible future changes.